### PR TITLE
Add test to nic

### LINKED ISSUE #33: Test NIC creation
Example:

```
	nic <= azure_nic_new($VM_NIC_NAME, $RESOURCE_GROUP_NAME, $LOCATION)
	nic <= azure_nic_set_vnet($nic, $VNET_NAME)
	nic <= azure_nic_set_subnet($nic, $SUBNET_PUBLIC_NAME)
	nic <= azure_nic_set_ipfw($nic, "true")
	nic <= azure_nic_set_publicip($nic, $VM_PUBLIC_IP_NAME)
	nic <= azure_nic_set_privateip($nic, $VM_ADDRESS)
	azure_nic_create($nic)
```

### PR DESCRIPTION
**Pull request by matheusvill**
Closes #33

### CODE REVIEW

File: tests/azure/nic_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	addrnic := "10.116.1.100"
+	subnetAddress := "10.116.1.0/24"
+
+	f.Shell.Run(
+		"./testdata/create_nsg.sh",
+		nsg,
+		f.ResGroupName,
+		f.Location,
+	)
+	_ = azure.NewNsg(f)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: Why this ?

File: tests/azure/nic_test.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+	f.Shell.Run(
+		"./testdata/create_vnet.sh",
+		vnet,
+		f.ResGroupName,
+		f.Location,
+		vnetAddress,
+	)
+	_ = azure.NewVnet(f)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: Why this ? (2)

File: tests/azure/nic_test.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	f.Shell.Run(
+		"./testdata/create_subnet.sh",
+		subnet,
+		f.ResGroupName,
+		vnet,
+		subnetAddress,
+		nsg,
+	)
+	_ = azure.NewSubnet(f)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: Why this ? (3)

Also, in Go, if you want to ignore the return values you can just call the function and not store any result at all, it works ;-)

File: tests/azure/subnet_test.go [hunk 4]
```diff
@@ -19,15 +19,31 @@ func testSubnetCreate(t *testing.T, f fixture.F) {
...
 	subnetAddress := "10.116.1.0/24"
+
+	f.Shell.Run(
+		"./testdata/create_vnet.sh",
+		vnet,
+		f.ResGroupName,
+		f.Location,
+		vnetAddress,
+	)
+	_ = azure.NewVnet(f)
```
Comment (katcipis) [inline tests/azure/subnet_test.go]: +1

File: tests/azure/subnet_test.go [hunk 5]
```diff
@@ -19,15 +19,31 @@ func testSubnetCreate(t *testing.T, f fixture.F) {
...
+
+	f.Shell.Run(
+		"./testdata/create_nsg.sh",
+		nsg,
+		f.ResGroupName,
+		f.Location,
+	)
+	_ = azure.NewNsg(f)
```
Comment (katcipis) [inline tests/azure/subnet_test.go]: +1

File: tests/azure/nic_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		f.ResGroupName,
+		nic,
+		f.Location,
+		vnet,
+		subnet,
+		addrnic,
+	)
+	nics := azure.NewNic(f)
+
+	nics.AssertExists(t, nic, nsg, addrnic)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: You configured the vnet/subnet/addrnic on the nic, it is enough to just validate the addnic ? Perhaps it is, since the addr is already on the subnet/vnet. Although, I think it is possible to have a different VNET with the same subnets/addrs, so a bug could slip in here (but it would be hard)

File: tests/lib/azure/nic.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		}
+		ip := *properties.IPConfigurations
+
+		if len(ip) == 0 || ip[0].PrivateIPAddress == nil {
+			return errors.New("The field AddressPrefix is nil!")
+		}
+
+		privateAddress := *ip[0].PrivateIPAddress
+		if privateAddress != address {
+			return errors.New("Subnet created with wrong Address. Expected: " + address + "Actual: " + privateAddress)
```
Comment (kamilash) [inline tests/lib/azure/nic.go]: Subnet? :)

File: tests/lib/azure/nic.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+			return err
+		}
+
+		if n.InterfacePropertiesFormat == nil {
+			return errors.New("The field InterfacePropertiesFormat is nil!")
+		}
+		properties := *n.InterfacePropertiesFormat
+
+		if properties.IPConfigurations == nil {
+			return errors.New("The field AddressPrefix is nil!")
```
Comment (kamilash) [inline tests/lib/azure/nic.go]: AddressPrefix?

File: tests/lib/azure/nic.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+		}
+		ip := *properties.IPConfigurations
+
+		if len(ip) == 0 || ip[0].PrivateIPAddress == nil {
+			return errors.New("The field AddressPrefix is nil!")
```
Comment (kamilash) [inline tests/lib/azure/nic.go]: AddressPrefix?

File: tests/azure/nic_test.go [hunk 10]
```diff
@@ -28,7 +28,7 @@ func testNicCreate(t *testing.T, f fixture.F) {
 		f.ResGroupName,
 		f.Location,
 	)
-	_ = azure.NewNsg(f)
+	azure.NewNsg(f)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: I was hinting that you dont need to call any New* methods if you are not going to use the return value. New functions are usually constructors, they dont have side effects (or should not at least), just calling it and ignoring the return is odd, and in this case I checked out the New functions and they do almost nothing. No need to call them here ;-)

Comment (katcipis) [inline tests/azure/nic_test.go]: Also no need to call any other of the New functions that the return value are ignored, they are basically the same (only the client changes)

File: tests/azure/nic_test.go (hunk 6 above)
Comment (matheusvill) [inline tests/azure/nic_test.go]: I didn't check the other things because if the other resources like vnet and subnet are wrong the nic won't be created. In the sample above I set the address to IP "10.116.3.100" and the subnet is "10.116.1.0/24":


11:39:48 info:    Executing command network nic create
11:39:48 info:    Looking up the network interface "klb-nic-tests-30"
11:39:49 info:    Looking up the subnet "klb-subnet-tests-91"
11:39:50 info:    Creating network interface "klb-nic-tests-30"
11:39:55 error:   Private static IP address 10.116.3.100 does not belong to the range of subnet prefix 10.116.1.0/24.
11:39:55 error:   Error information has been recorded to /home/matheus/.azure/azure.err
11:39:55 error:   network nic create command failed

File: tests/azure/nic_test.go (hunk 3 above)
Comment (matheusvill) [inline tests/azure/nic_test.go]: My fault only a bad habit, fixed! :+1:

File: tests/lib/azure/nic.go (hunk 8 above)
Comment (matheusvill) [inline tests/lib/azure/nic.go]: Fixed :+1:

File: tests/azure/nic_test.go (hunk 6 above)
Comment (katcipis) [inline tests/azure/nic_test.go]: Since the resource group isolates from all the rest and we always build new ones, it makes sense to ignore on this case. Without the isolation of the resource group you could have a bug where klb choses the vnet wrongly (you can have vnets with the same network CIDR ... I think). Anyway, really does not seem like it is a requirement here

File: tests/azure/nic_test.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	addrnic := "10.116.1.100"
+	subnetAddress := "10.116.1.0/24"
+
+	f.Shell.Run(
+		"./testdata/create_nsg.sh",
+		nsg,
+		f.ResGroupName,
+		f.Location,
+	)
+	azure.NewNsg(f)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: I think github eated my comments about this :-(. These calls to New* are not required, the constructor does nothing. Why is this here ?

File: tests/azure/nic_test.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+	f.Shell.Run(
+		"./testdata/create_vnet.sh",
+		vnet,
+		f.ResGroupName,
+		f.Location,
+		vnetAddress,
+	)
+	azure.NewVnet(f)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: Not necessary +1

File: tests/azure/nic_test.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+	f.Shell.Run(
+		"./testdata/create_subnet.sh",
+		subnet,
+		f.ResGroupName,
+		vnet,
+		subnetAddress,
+		nsg,
+	)
+	azure.NewSubnet(f)
```
Comment (katcipis) [inline tests/azure/nic_test.go]: Not necessary +1

File: tests/azure/subnet_test.go [hunk 14]
```diff
@@ -19,15 +19,31 @@ func testSubnetCreate(t *testing.T, f fixture.F) {
...
 	subnetAddress := "10.116.1.0/24"
+
+	f.Shell.Run(
+		"./testdata/create_vnet.sh",
+		vnet,
+		f.ResGroupName,
+		f.Location,
+		vnetAddress,
+	)
+	azure.NewVnet(f)
```
Comment (katcipis) [inline tests/azure/subnet_test.go]: Not necessary +1

File: tests/azure/subnet_test.go [hunk 15]
```diff
@@ -19,15 +19,31 @@ func testSubnetCreate(t *testing.T, f fixture.F) {
...
+
+	f.Shell.Run(
+		"./testdata/create_nsg.sh",
+		nsg,
+		f.ResGroupName,
+		f.Location,
+	)
+	azure.NewNsg(f)
```
Comment (katcipis) [inline tests/azure/subnet_test.go]: Not necessary +1
